algorithms/a3c/env_vizdoom.py

```diff
@@ -0,0 +1,379 @@
+
+        output_objects = []
+
+        # For instructions like "largest red object", the incorrect object
+        # set could contain larger objects of other color
+
+        for obj in self.objects_info:
+            if instr_contains_color:
+                # first check color attribute
+                if correct_object.color != obj.color:
+                    output_objects.append(obj)
+
+            if instruction_words[3] == 'largest':
+                if correct_object.absolute_size > \
+                        obj.absolute_size + SIZE_THRESHOLD:
+                    output_objects.append(obj)
+
+            else:
+                if correct_object.absolute_size < \
+                        obj.absolute_size - SIZE_THRESHOLD:
+                    output_objects.append(obj)
+
+        # shuffle the objects and select the top 4
+        # randomizing the objects combination
+        random.shuffle(output_objects)
+        return [x.name for x in output_objects[:4]]
+
+    def get_objects_info(self):
+        objects = []
+        objects_map = self.objects
+        for obj in objects_map:
+            split_word = split_object(obj)
+            candidate_object = DoomObject(*split_word)
+            candidate_object.absolute_size = self.object_sizes[obj]
+            objects.append(candidate_object)
+
+        return objects
+
+    def get_all_objects(self, filename):
+        objects = []
+        object_dict = {}
+        count = 0
+        instructions = self.get_instr(filename)
+        for instruction_data in instructions:
+            object_names = instruction_data['targets']
+            for object_name in object_names:
+                if object_name not in objects:
+                    objects.append(object_name)
+                    object_dict[object_name] = count
+                    count += 1
+
+        return objects, object_dict
+
+    def get_target_objects(self, instr_id):
+        object_names = self.instructions[instr_id]['targets']
+        correct_objects = []
+        for object_name in object_names:
+            correct_objects.append(self.object_dict[object_name])
+
+        return correct_objects
+
+    def get_instr(self, filename):
+        with open(filename, 'rb') as f:
+            instructions = json.load(f)
+        return instructions
+
+    def read_size_file(self, filename):
+        with codecs.open(filename, 'r') as open_file:
+            lines = open_file.readlines()
+
+        object_sizes = {}
+        for i, line in enumerate(lines):
+            split_line = line.split('\t')
+            if split_line[1].strip() in self.objects:
+                object_sizes[split_line[1].strip()] = int(split_line[2])
+
+        return object_sizes
+
+    def get_random_instruction(self):
+        instruction_id = np.random.randint(len(self.instructions))
+        instruction = self.instructions[instruction_id]['instruction']
+
+        return instruction, instruction_id
+
+    def get_word_to_idx(self):
```

Comment:
I don't understand. word_to_idx and idx_to_word are the two most common NLP coding patterns.

Reply:
If you want to get elements by index or name a list is easier to use than a dictionary. 
My suggestion is changing this `word_to_idx[word] = len(word_to_idx)` to this `word_to_idx.append(word)` (of course the object definition has to be changed as well). Isn't that obviously better? Why do you need keys and values instead? 
Since what I am telling you is the standard in the python world, could you please argue why something different is needed here?

Reply:
have you seen NLP code before?

we need to go from word to idx, not idx to word. The list works only in the latter but not the former case

How else are you going to find the word embedding index from just a string?